### PR TITLE
Use mir constant in thir instead of ty::Const

### PR DESCRIPTION
**Pull request by b-naber**
This is blocked on https://github.com/rust-lang/rust/pull/94059 (does include its changes, the first two commits in this PR correspond to those changes) and https://github.com/rust-lang/rust/pull/93800 being reinstated (which had to be reverted). Mainly opening since @lcnr offered to give some feedback and maybe also for a perf-run (if necessary).

This currently contains a lot of duplication since some of the logic of `ty::Const` had to be copied to `mir::ConstantKind`, but with the introduction of valtrees a lot of that functionality will disappear from `ty::Const`.

Only the last commit contains changes that need to be reviewed here. Did leave some `FIXME` comments regarding future implementation decisions and some things that might be incorrectly implemented.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/const_evaluatable.rs [hunk 1]
```diff
@@ -200,14 +200,17 @@ impl<'tcx> AbstractConst<'tcx> {
         Ok(inner.map(|inner| AbstractConst { inner, substs: uv.substs }))
     }
 
-    pub fn from_const(
+    pub fn from_constant(
```
Comment (BoxyUwU) [inline compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:203]: `ConstantKind::Value` is for non type system stuff isnt it? all of the stuff in this file is to do with type level constants so it seems kinda sus to be taking a `ConstantKind` here instead of a `Const`. I'm not all that familiar with `mir::Constant` so I may be off here :thinking:

Comment (b-naber) [reply compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:203]: Yes `ConstantKind::Value` isn't meant to be seen by the type system, but I don't think this is a problem here since we basically discard those here and only care about unevaluated constants. I think this is unlikely to change given the purpose of `AbstractConst`.

**Review by lcnr — COMMENTED** [review]
will get to the rest of this PR next week. We originally intended to also use valtrees in pattern, as they should simplify exhaustiveness checking.

Currently trying to figure out what exactly our plan should be for that and more generally how to deal with structural equality going forward

File: compiler/rustc_middle/src/mir/mod.rs [hunk 2]
```diff
@@ -2623,6 +2666,183 @@ impl<'tcx> ConstantKind<'tcx> {
...
+        Self::Val(cv, ty)
+    }
+
+    pub fn from_usize(tcx: TyCtxt<'tcx>, n: u64) -> Self {
+        let ty = tcx.types.usize;
+        Self::from_bits(tcx, n as u128, ty::ParamEnv::empty().and(ty))
+    }
+
+    #[instrument(skip(tcx), level = "debug")]
+    pub fn try_eval_lit_or_param(
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_middle/src/mir/mod.rs:2702]: Mir constants do not need eager normalization per se, but it can be helpful for perf if mir opts can trust that they don't even need to attempt to eval them, because all evaluable ones have been evaluated

**Comment by oli-obk** [comment]
> . We originally intended to also use valtrees in pattern, as they should simplify exhaustiveness checking.

Yea. My plan was "evaluate to valtree and if that fails, keep the constant opaque". I'm not sure we can fall back to "always be opaque" without a breaking change. And keeping around the deref/field infrastructure for mir constants is a cost I'd rather avoid.

File: compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs [hunk 3]
```diff
@@ -630,9 +640,9 @@ pub(super) enum Constructor<'tcx> {
     /// Ranges of integer literal values (`2`, `2..=5` or `2..5`).
     IntRange(IntRange),
     /// Ranges of floating-point literal values (`2.0..=5.2`).
-    FloatRange(ty::Const<'tcx>, ty::Const<'tcx>, RangeEnd),
+    FloatRange(mir::ConstantKind<'tcx>, mir::ConstantKind<'tcx>, RangeEnd),
```
Comment (lcnr) [inline compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs:643]: :thinking: i guess using mir constant here is fine :thinking: float ranges are weird anyways

File: compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs [hunk 4]
```diff
@@ -630,9 +640,9 @@ pub(super) enum Constructor<'tcx> {
...
     /// String literals. Strings are not quite the same as `&[u8]` so we treat them separately.
-    Str(ty::Const<'tcx>),
+    Str(mir::ConstantKind<'tcx>),
```
Comment (lcnr) [inline compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs:645]: as we structurally match strings, we should use `ty::Const<'tcx>` here

Comment (lcnr) [reply compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs:645]: or tbh, probably we should use `Str(Valtree)` :sweat_smile: not sure what's the best way to do this

File: compiler/rustc_trait_selection/src/traits/const_evaluatable.rs (hunk 1 above)
Comment (lcnr) [reply compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:203]: we don't want this, should never need an `AbstractConst` for mir constants. These are only used to unify constants in the type system

File: compiler/rustc_trait_selection/src/traits/const_evaluatable.rs [hunk 5]
```diff
@@ -294,7 +297,7 @@ impl<'a, 'tcx> AbstractConstBuilder<'a, 'tcx> {
                 }
             }
 
-            fn visit_const(&mut self, ct: ty::Const<'tcx>) {
+            fn visit_constant(&mut self, ct: mir::ConstantKind<'tcx>) {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:300]: probably need both visitors here, both mir constants and ty constants.

File: compiler/rustc_middle/src/thir/abstract_const.rs [hunk 6]
```diff
@@ -22,7 +22,7 @@ pub enum CastKind {
 /// A node of an `AbstractConst`.
 #[derive(Debug, Clone, Copy, PartialEq, Eq, HashStable, TyEncodable, TyDecodable)]
 pub enum Node<'tcx> {
-    Leaf(ty::Const<'tcx>),
+    Leaf(mir::ConstantKind<'tcx>),
```
Comment (lcnr) [inline compiler/rustc_middle/src/thir/abstract_const.rs:25]: Leaf should keep being a `ty::Const`. When encountering a mir constant which can't be converted to a valtree, we should error.

**Review by lcnr — COMMENTED** [review]
Changing the thir stuff partially back to `ty::Const` isn't strictly necessary and is probably easier once valtrees are fully set up.

Changing `Node::Leaf` back to `ty::Const` is important to me as these are constants we use in the type system

File: compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs (hunk 4 above)
Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs:645]: `&'tcx [u8]` would be the least ambiguous, but may require some conversions or interning later. Right now you can get it right from the `Allocation`

File: compiler/rustc_const_eval/src/const_eval/mod.rs [hunk 7]
```diff
@@ -207,3 +241,39 @@ pub(crate) fn deref_const<'tcx>(
 
     tcx.mk_const(ty::ConstS { val: ty::ConstKind::Value(op_to_const(&ecx, &mplace.into())), ty })
 }
+
+#[instrument(skip(tcx), level = "debug")]
+pub(crate) fn deref_mir_constant<'tcx>(
```
Comment (lcnr) [inline compiler/rustc_const_eval/src/const_eval/mod.rs:246]: same here, `deref_mir_constant` doesn't seem to be used anywhere

File: compiler/rustc_middle/src/mir/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lcnr) [reply compiler/rustc_middle/src/mir/mod.rs:2784]: can you remove this function and instead eagerly call `tcx.const_eval_resolve` on before building mir constants?

File: compiler/rustc_middle/src/thir/visit.rs [hunk 8]
```diff
@@ -24,6 +25,8 @@ pub trait Visitor<'a, 'tcx: 'a>: Sized {
     fn visit_pat(&mut self, pat: &Pat<'tcx>) {
         walk_pat(self, pat);
     }
+
+    fn visit_constant(&mut self, _constant: ConstantKind<'tcx>) {}
```
Comment (lcnr) [inline compiler/rustc_middle/src/thir/visit.rs:29]: this looks like dead code to me

File: compiler/rustc_mir_build/src/thir/cx/mod.rs [hunk 9]
```diff
@@ -75,6 +78,24 @@ impl<'tcx> Cx<'tcx> {
         }
     }
 
+    #[instrument(skip(self), level = "debug")]
+    crate fn const_eval_literal(
```
Comment (lcnr) [inline compiler/rustc_mir_build/src/thir/cx/mod.rs:82]: that function seems to be unused? :thinking:

File: compiler/rustc_mir_build/src/thir/pattern/mod.rs [hunk 10]
```diff
@@ -739,15 +739,15 @@ impl<'tcx> PatternFoldable<'tcx> for PatKind<'tcx> {
     }
 }
 
+// FIXME: Get rid of this function once valtrees land
```
Comment (lcnr) [inline compiler/rustc_mir_build/src/thir/pattern/mod.rs:742]: I don't think we can remove this function, though we probably want to stop using `ty::Const` here.
```suggestion
```

File: compiler/rustc_mir_build/src/thir/pattern/mod.rs [hunk 11]
```diff
@@ -759,7 +759,8 @@ crate fn compare_const_vals<'tcx>(
 
     // Early return for equal constants (so e.g. references to ZSTs can be compared, even if they
     // are just integer addresses).
-    if a.val() == b.val() {
+    // FIXME This might be wrong
+    if a == b {
```
Comment (lcnr) [inline compiler/rustc_mir_build/src/thir/pattern/mod.rs:763]: why this change?

File: compiler/rustc_middle/src/thir/visit.rs (hunk 8 above)
Comment (b-naber) [reply compiler/rustc_middle/src/thir/visit.rs:29]: We need this in the `is_poly` visitor in abstract-const creation.

Comment (lcnr) [reply compiler/rustc_middle/src/thir/visit.rs:29]: Yes, my idea is to directly check that in `visit_expr` or `visit_pat` instead of using `visit_constant`.

Considering that `visit_constant` doesn't actually visit most constants in the body, that method seems to me to pretty much only cause confusion

**Review by oli-obk — APPROVED** [review]

**Comment by b-naber** [comment]
I don't think that we actually need `lit_to_mir_constant` to be a constant anymore. Before reverting pattern matching back to `ty::Const`,  it would have probably made sense to have a query for`lit_to_mir_constant`, given that `lit_to_const` is also a query, but now we only use it in `as_constant`.

**Comment by pnkfelix** [comment]
Visiting for weekly performance triage. This caused a 0.2% regression on the primary benchmarks webrender-2022 check (full) and cargo-0.60.0 check (incr-full). An [earlier perf run](https://github.com/rust-lang/rust/pull/94255#issuecomment-1093347875), on I think the same commit series as what was eventually merged, showed no primary regressions at all.

A number of secondary benchmarks regressed, but they are all stress test microbenchmarks, and should not block landing work of this nature.

So, I'm chalking the minor amount of regression that was observed up to noise.

@rustbot label +perf-regression-triaged
